containers/runtime/xds/config/config.go

```diff
@@ -0,0 +1,436 @@
+			log.Fatalf("fail to unmarshal %v resource: %v", resource.SecretType, err)
+		}
+		if err := parsedSecret.ValidateAll(); err != nil {
+			log.Fatalf("failed to validate the parsed %v: %v", resource.SecretType, err)
+		}
+		cr.Resource = &parsedSecret
+	case resource.ExtensionConfigType:
+		parsedExtensionConfig := extension.EcdsDummy{}
+		if err := anypb.UnmarshalTo(&a, &parsedExtensionConfig, proto.UnmarshalOptions{}); err != nil {
+			log.Fatalf("fail to unmarshal %v resource: %v", resource.ExtensionConfigType, err)
+		}
+		if err := parsedExtensionConfig.ValidateAll(); err != nil {
+			log.Fatalf("failed to validate the parsed %v: %v", resource.ExtensionConfigType, err)
+		}
+		cr.Resource = &parsedExtensionConfig
+	}
+	return nil
+}
+
+// TestResource provides the names of the resources.
+type TestResource struct {
+	// xDSServerClusterName is the xDS server's name supplied to Envoy bootstrap file.
+	// This field is needed if the RouteConfig or EdsConfig are fetched from the
+	// xDS server separately.
+	XDSServerClusterName string
+	// testServiceClusterName is the name of the upstream clsuter,
+	// consisted by the test servers.
+	TestServiceClusterName string
+	// testRouteName is the route name returned by the test listeners.
+	TestRouteName string
+	// testGrpcListenerName is only used by gRPC xDS client, must be used as xds:///<listener_name>,
+	// API listener is required.
+	TestGrpcListenerName string
+	// testEnvoyListenerName is only used by Envoy proxy.
+	TestEnvoyListenerName string
+	// testListenerPort is only used by Envoy, socket listener, traffic will be directly direct to
+	// this port.
+	TestListenerPort uint32
+	// testEndpointName is the name of the cluster. This will be the `service_name
+	// in <envoy_v3_api_field_config.cluster.v3.Cluster.EdsClusterConfig.service_name>` value if specified
+	// in the cluster.
+	TestEndpointName string
+	// List of endpoints to load balance to, these will all goes into the same Endpoint resource.
+	TestEndpoints []*TestEndpoint
+}
+
+// TestEndpoint is the address and the port of the backends.
+type TestEndpoint struct {
+	// TestUpstreamHost is upstream host address
+	TestUpstreamHost string
+	// TestUpstreamHost is upstream host port
+	TestUpstreamPort uint32
+}
+
+func (t *TestResource) validateResource(snap cache.Snapshot) error {
+	// gRPC listener name must match match the server_target_string in xds:///server_target_string"
+	listenerType := cache.GetResponseType(resource.ListenerType)
+	if _, ok := snap.Resources[listenerType].Items[t.TestGrpcListenerName]; !ok {
+		log.Fatalf("failed validation of listener resource, please set up gRPC listener with name %v", t.TestGrpcListenerName)
+	}
+
+	// Envoy listener's port value is used by routing traffic to Envoy sidecar
+	for listenerName, listenerWithTTL := range snap.Resources[listenerType].Items {
+		// listener port is not required for gRPC listener.
+		if listenerName != t.TestGrpcListenerName {
```

Comment:
I'm not sure what you mean by "gRPC listener" here.  Do you mean `api_listener`s? If so, this check should probably be made more general so that all `api_listener`s are excluded.

Reply:
You are right, the gRPC listeners is just a generic api_listener. Seems like the api_listener is only used for non-proxy listeners. Update it to be more generic. 
https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/listener/v3/listener.proto

Reply:
Even we skipped port checking for the api_listener, we still need to add the port field in the default config to pass the listener validation check, it should not affect the api_listener, but will double check to verify.

https://source.corp.google.com/piper///depot/google3/third_party/envoy/src/test/server/listener_manager_impl_test.cc;l=4762?q=api_listener

Reply:
> Seems like the api_listener is only used for non-proxy listeners
Actually, the opposite is true. `api_listeners` are queried by proxyless clients.

> Even we skipped port checking for the api_listener, we still need to add the port field in the default config to pass the listener validation check, it should not affect the api_listener, but will double check to verify.

SGTM